rosys/hardware/lizard_firmware.py

```diff
@@ -105,13 +105,20 @@ async def flash_core(self) -> None:
 
     async def flash_p0(self) -> None:
         rosys.notify(f'Flashing Lizard firmware {self.core_version} to P0...')
+        flash_p0_complete = False
+
+        def flash_callback():
+            nonlocal flash_p0_complete
+            flash_p0_complete = True
+        self.robot_brain.FLASH_P0_COMPLETE.register(flash_callback)
```

Comment:
This could be simplified like this:
```py
state = {'complete': False}
self.robot_brain.FLASH_P0_COMPLETE.register(lambda: state.update(complete=True))
```
A dictionary seems a bit like overkill. But it allows to squeeze all code into two lines, which - in my opinion - improves readability in this case.

BUT we need to unregister the callback later, which won't work for a lambda statement. Therefore I'd suggest to introduce a member variable in the initializer that is always updated when the event is received:
```py
self._p0_flash = {'last_complete': 0}
self.robot_brain.FLASH_P0_COMPLETE.register(lambda: self._p0_flash.update(last_complete=rosys.time()))
```
Then you can simply check
```py
while self._p0_flash['last_complete'] < start:
    ...
```

Reply:
> BUT we need to unregister the callback later, which won't work for a lambda statement.

Yeah, this is why I choose that helper function, but your solution is the better way.

We could use setattr to avoid a dictionary, but I'm not sure if this really helps with readability
```py
self.last_p0_flash_complete = 0
self.robot_brain.FLASH_P0_COMPLETE.register(lambda: setattr(self, 'last_p0_flash_complete', rosys.time()))
```